### PR TITLE
Run tinode only if init-db returns success.

### PR DESCRIPTION
**Pull request by aforge**
In tinode:tinode docker:
1. Check stderr of init-db for the presence of `All done` string.
2. If the string isn't found, stop. Otherwise, proceed to running tinode.
3. Use `SAMPLE_DATA` env var to specify the sample data file.

### CODE REVIEW

File: docker/tinode/Dockerfile [hunk 1]
```diff
@@ -102,6 +102,7 @@ RUN tar -xzf tinode-$TARGET_DB.linux-amd64.tar.gz \
 	&& rm tinode-$TARGET_DB.linux-amd64.tar.gz
 
 # Copy config template to the container.
```
Comment (or-else) [inline docker/tinode/Dockerfile]: This comment should be one line lower.

File: docker/tinode/Dockerfile [hunk 2]
```diff
@@ -102,6 +102,7 @@ RUN tar -xzf tinode-$TARGET_DB.linux-amd64.tar.gz \
...
+COPY init-db .
```
Comment (or-else) [inline docker/tinode/Dockerfile]: I believe this line copies init-db from your host docker directory to container. You should already have it in your container after downloading and unpacking `tinode-$TARGET_DB.linux-amd64.tar.gz`.

File: docker/tinode/entrypoint.sh [hunk 3]
```diff
@@ -50,8 +50,8 @@ else
 fi
 
 # Load default sample data when generating or resetting the database.
```
Comment (or-else) [inline docker/tinode/entrypoint.sh]: Update comment?

File: docker/tinode/entrypoint.sh [hunk 4]
```diff
@@ -89,42 +89,32 @@ if [ ! -z "$IOS_UNIV_LINKS_APP_ID" ] ; then
...
+# If sample data was provided, try to find Tino password.
+if [ ! -z "$SAMPLE_DATA" ] ; then
+	grep "usr;tino;" $init_stdout > /botdata/tino-password
 fi
 
-if [ "$run_tinode" != "true" ]; then
-  # If we don't want to run tinode, we are done.
-  exit 0
+# Check if the init-db output contains the magic string.
+grep -q "All done" $init_stderr
```
Comment (or-else) [inline docker/tinode/entrypoint.sh]: Please consider checking for exit code:
```
if [ $? -eq 0 ]
then
  echo "Success"
else
  echo "Failure" >&2
fi
```

File: tinode-db/main.go [hunk 5]
```diff
@@ -225,11 +225,20 @@ func main() {
 		// Upgrade DB from one version to another.
 		err = store.UpgradeDb(config.StoreConfig)
 		if err == nil {
-			log.Println("Database successfully upgraded")
+			log.Println("Database successfully upgraded. All done.")
```
Comment (or-else) [inline tinode-db/main.go]: Maybe `os.Exit(0)` and `os.Exit(1)` would be more idiomatic.

**Comment by or-else** [comment]
> Check stderr of init-db for the presence of All done string.

If we have to modify init-db anyway, `os.Exit()` would be a better solution.

File: tinode-db/main.go (hunk 5 above)
Comment (aforge) [reply tinode-db/main.go]: Do you mean checking for the exit code instead of the magic string?

File: docker/tinode/entrypoint.sh (hunk 3 above)
Comment (aforge) [reply docker/tinode/entrypoint.sh]: Done.

File: docker/tinode/entrypoint.sh (hunk 4 above)
Comment (aforge) [reply docker/tinode/entrypoint.sh]: Done below.

File: docker/tinode/Dockerfile (hunk 2 above)
Comment (aforge) [reply docker/tinode/Dockerfile]: This line is not necessary. Removed.

File: docker/tinode/entrypoint.sh (hunk 4 above)
Comment (or-else) [reply docker/tinode/entrypoint.sh]: Then you don't need the `grep -q "All done" $init_stderr` line anymore.

File: tinode-db/main.go (hunk 5 above)
Comment (or-else) [reply tinode-db/main.go]: Yes. Sorry for not making myself clear.

Comment (aforge) [reply tinode-db/main.go]: Done.

File: docker/tinode/entrypoint.sh [hunk 6]
```diff
@@ -49,9 +49,9 @@ else
   STATIC_DIR="./static"
 fi
 
-# Load default sample data when generating or resetting the database.
-if [[ -z "$SAMPLE_DATA" && "$UPGRADE_DB" = "false" ]] ; then
-	SAMPLE_DATA="$DEFAULT_SAMPLE_DATA"
+# Do not load data when upgrading database.
+if [[ "$UPGRADE_DB" = "true" ]] ; then
+	SAMPLE_DATA=""
```
Comment (or-else) [inline docker/tinode/entrypoint.sh]: Just `SAMPLE_DATA=`

File: docker/tinode/entrypoint.sh [hunk 7]
```diff
@@ -89,42 +89,22 @@ if [ ! -z "$IOS_UNIV_LINKS_APP_ID" ] ; then
...
-  # Maybe load sample data?
-  if [ ! -z "$SAMPLE_DATA" ] ; then
-    init_args+=("--data=$SAMPLE_DATA")
-  fi
-  # Initialize the database if it has not been initialized yet or if data reset/upgrade has been requested.
-  ./init-db "${init_args[@]}" | grep "usr;tino;" > /botdata/tino-password
+init_args=("--reset=${RESET_DB}" "--upgrade=${UPGRADE_DB}" "--config=${CONFIG}")
+# Maybe load sample data?
+if [ ! -z "$SAMPLE_DATA" ] ; then
+	init_args+=("--data=$SAMPLE_DATA")
```
Comment (or-else) [inline docker/tinode/entrypoint.sh]: I don't think this check is needed. Passing `--data=` should work just as well.

File: docker/tinode/entrypoint.sh [hunk 8]
```diff
@@ -140,10 +120,10 @@ args=("--config=${CONFIG}" "--static_data=$STATIC_DIR")
 
 # Maybe set node name in the cluster.
 if [ ! -z "$CLUSTER_SELF" ] ; then
-  args+=("--cluster_self=$CLUSTER_SELF")
+	args+=("--cluster_self=$CLUSTER_SELF")
```
Comment (or-else) [inline docker/tinode/entrypoint.sh]: The check is redundant. Empty parameter `--cluster_self=` is equivalent to no parameter.

File: docker/tinode/entrypoint.sh [hunk 9]
```diff
@@ -140,10 +120,10 @@ args=("--config=${CONFIG}" "--static_data=$STATIC_DIR")
...
 fi
 if [ ! -z "$PPROF_URL" ] ; then
-  args+=("--pprof_url=$PPROF_URL")
+	args+=("--pprof_url=$PPROF_URL")
```
Comment (or-else) [inline docker/tinode/entrypoint.sh]: Likewise, redundant check.

File: docker/tinode/entrypoint.sh [hunk 10]
```diff
@@ -140,10 +120,10 @@ args=("--config=${CONFIG}" "--static_data=$STATIC_DIR")
...
 fi
 
 # Create the log directory (/var/log/tinode-`current timestamp`).
```
Comment (or-else) [inline docker/tinode/entrypoint.sh]: I'm not sure we need to setup log rotation here. I would just write to log file and be done with it. Users may have their own logging needs and methods.

File: tinode-db/main.go [hunk 11]
```diff
@@ -218,18 +218,27 @@ func main() {
 		log.Println("Database reset requested")
 	} else {
 		log.Println("Database exists, DB version is correct. All done.")
-		return
+		os.Exit(0)
```
Comment (or-else) [inline tinode-db/main.go]: Also add `os.Exit(0)` as the very last line of this file, after `genDb(&data)`. Github does not let me add a comment there.

Comment (aforge) [reply tinode-db/main.go]: Done.

File: docker/tinode/entrypoint.sh (hunk 9 above)
Comment (aforge) [reply docker/tinode/entrypoint.sh]: Done.

File: docker/tinode/entrypoint.sh (hunk 10 above)
Comment (aforge) [reply docker/tinode/entrypoint.sh]: Removed.

**Pull request merged**